neqo-transport/src/connection.rs

```diff
@@ -292,17 +298,36 @@ impl StateSignaling {
         }
     }
 
-    pub fn closing(&self) -> bool {
-        *self == Self::ConnectionClose
+    pub fn close<S>(&mut self, error: ConnectionError, frame_type: FrameType, message: S)
+    where
+        S: AsRef<str>,
+    {
+        let msg: &[u8] = message.as_ref().as_ref();
+        let frame = Frame::ConnectionClose {
+            error_code: CloseError::from(error),
+            frame_type,
+            reason_phrase: msg.to_vec(),
+        };
+        *self = Self::ConnectionClose(frame);
     }
 
-    pub fn close(&mut self) {
-        *self = Self::ConnectionClose
+    /// If a close is pending, take it.
+    pub fn close_frame(&mut self) -> Option<Frame> {
+        if let Self::ConnectionClose(frame) = self {
+            let frame = mem::replace(frame, Frame::Padding);
+            *self = Self::CloseSent(frame.clone());
+            Some(frame)
```

Comment:
Maybe try defining ConnectionClose as ConnectionClose(Option<Frame>) then use `Option::take()`? Less... weird. Same for `CloseSent` below.

That might also let you define data members in ConnectionClose as the data in a close frame rather than an actual `Frame`, then build the Close frame here? So there would be no possibility of another frame type slipping in there somehow. Dunno, just an idea.

Reply:
This one is not really something that `Option::take()` can support.  Because we aren't really taking the value.  We need to keep the value for later, in case we need to send another.  That said, this isn't 100% right now that I've thought this through again.  (This hasn't captured the difference between closing and draining properly.